File: packages/tui/src/form/builder.ts
import { alert, datatable, error, info, intro, note, outro, table, warning } from '#tui/output';
import { autocomplete, confirm, multiselect, multisearch, pause, search, select, suggest } from '#tui/prompts/choices';
import { number, password, text, textarea } from '#tui/prompts/basic';
import { progress, spin, stream, task } from '#tui/status';
import type { FormResponses, FormStep } from '#tui/form/types';
import type { ChoiceInput, DataTablePromptOptions, MaybePromise, MultiSearchPromptOptions, SearchPromptOptions, TableOptions, TextPromptOptions } from '#tui/types';
import type { SuggestOptions } from '#tui/prompts/choices';
import type { Logger, Progress } from '#tui/status';

export class FormBuilder {
	readonly #steps: FormStep[] = [];
	readonly #responses: FormResponses = [] as unknown as FormResponses;

	add(step: (responses: FormResponses, previous: unknown) => MaybePromise<unknown>, name?: string, ignoreWhenReverting = false): this {
		this.#steps.push({
			condition: true,
			ignoreWhenReverting,
			name,
			run: step,
		});

		return this;
	}

	addIf(
		condition: boolean | ((responses: FormResponses) => boolean),
		step: (responses: FormResponses, previous: unknown) => MaybePromise<unknown>,
		name?: string,
		ignoreWhenReverting = false,
	): this {
		this.#steps.push({
			condition,
			ignoreWhenReverting,
			name,
			run: step,
		});

		return this;
	}

	async submit(): Promise<FormResponses> {
		for (const [index, step] of this.#steps.entries()) {
			const key = step.name ?? index;
			const shouldRun = typeof step.condition === 'function' ? step.condition(this.#responses) : step.condition;

			if (!shouldRun) {
				this.#responses[key] = null;
				continue;
			}

			this.#responses[key] = await step.run(this.#responses, this.#responses[key]);
		}

		return this.#responses;
	}

	text(
		label: string,
		placeholder = '',
		defaultValue = '',
		required: boolean | string = false,
		validate: TextPromptOptions['validate'] = undefined,
		hint = '',
		name?: string,
		transform?: TextPromptOptions['transform'],
	): this {
		return this.add((_, previous) => text(label, placeholder, previous === undefined || previous === null ? defaultValue : String(previous), required, validate, hint, transform), name);
	}

	textarea(
		label: string,
		placeholder = '',
		defaultValue = '',
		required: boolean | string = false,
		validate: TextPromptOptions['validate'] = undefined,
		hint = '',
		rows = 5,
		name?: string,
		transform?: TextPromptOptions['transform'],
	): this {
		return this.add((_, previous) => textarea(label, placeholder, previous === undefined || previous === null ? defaultValue : String(previous), required, validate, hint, rows, transform), name);
	}

	password(
		label: string,
		placeholder = '',
		required: boolean | string = false,
		validate: TextPromptOptions['validate'] = undefined,
		hint = '',
		name?: string,
		transform?: TextPromptOptions['transform'],
	): this {
		return this.add(() => password(label, placeholder, required, validate, hint, transform), name);
	}

	number(
		label: string,
		placeholder = '',
		defaultValue = 0,
		required: boolean | string = false,
		validate: ((value: number) => MaybePromise<string | false | null | undefined>) | undefined = undefined,
		hint = '',
		min?: number,
		max?: number,
		step?: number,
		name?: string,
	): this {
		return this.add((_, previous) => number(label, placeholder, typeof previous === 'number' ? previous : defaultValue, required, validate, hint, min, max, step), name);
	}

	confirm(
		label: string,
		defaultValue = true,
		yes = 'Yes',
		no = 'No',
		required: boolean | string = false,
		validate: ((value: boolean) => MaybePromise<string | false | null | undefined>) | undefined = undefined,
		hint = '',
		name?: string,
	): this {
		return this.add(() => confirm(label, defaultValue, yes, no, required, validate, hint), name);
	}

	select<T>(
		label: string,
		options: Array<ChoiceInput<T>>,
		defaultValue?: T,
		scroll = 5,
		validate: ((value: T) => MaybePromise<string | false | null | undefined>) | undefined = undefined,
		hint = '',
		required: boolean | string = true,
		name?: string,
	): this {
		return this.add((_, previous) => select({ message: label, options, default: previous === undefined ? defaultValue : (previous as T), scroll, validate, hint, required }), name);
	}

	multiselect<T>(
		label: string,
		options: Array<ChoiceInput<T>>,
		defaultValue: T[] = [],
		scroll = 5,
		required: boolean | string = false,
		validate: ((value: T[]) => MaybePromise<string | false | null | undefined>) | undefined = undefined,
		hint = 'Use the space bar to select options.',
		name?: string,
	): this {
		return this.add((_, previous) => multiselect({ message: label, options, default: Array.isArray(previous) ? (previous as T[]) : defaultValue, scroll, required, validate, hint }), name);
	}

	suggest(
		label: string,
		options: SuggestOptions['options'],
		defaultValue = '',
		scroll = 5,
		required: boolean | string = false,
		validate: TextPromptOptions['validate'] = undefined,
		hint = '',
		name?: string,
		transform?: TextPromptOptions['transform'],
	): this {
		return this.add(
			(_, previous) =>
				suggest({ message: label, label, options, default: previous === undefined || previous === null ? defaultValue : String(previous), scroll, required, validate, hint, transform }),
			name,
		);
	}

	autocomplete(
		label: string,
		options: SuggestOptions['options'],
		defaultValue = '',
		required: boolean | string = false,
		validate: TextPromptOptions['validate'] = undefined,
		hint = '',
		name?: string,
		transform?: TextPromptOptions['transform'],
	): this {
		return this.add(
			(_, previous) =>
				autocomplete({ message: label, label, options, default: previous === undefined || previous === null ? defaultValue : String(previous), required, validate, hint, transform }),
			name,
		);
	}

	search<T>(options: SearchPromptOptions<T>, name?: string): this {
		return this.add(() => search(options), name);
	}

	multisearch<T>(options: MultiSearchPromptOptions<T>, name?: string): this {
		return this.add(() => multisearch(options), name);
	}

	spin<T>(callback: () => MaybePromise<T>, message = '', name?: string): this {
		return this.add(() => spin(callback, { message }), name, true);
	}

	task<T>(label: string, callback: (logger: Logger) => MaybePromise<T>, limit = 10, keepSummary = false, subLabel = '', name?: string): this {
		return this.add(() => task(label, callback, limit, keepSummary, subLabel), name);
	}

	pause(message = 'Press enter to continue', name?: string): this {
		return this.add(
			async () => {
				await pause(message);

				return null;
			},
			name,
			true,
		);
	}

	stream(source: AsyncIterable<string> | Iterable<string>, name?: string): this {
		return this.add(
			async () => {
				await stream(source);

				return null;
			},
			name,
			true,
		);
	}

	note(message: string, type: string | null = null, name?: string): this {
		return this.add(
			() => {
				note(message, type);

				return null;
			},
			name,
			true,
		);
	}

	error(message: string, name?: string): this {
		return this.add(
			() => {
				error(message);

				return null;
			},
			name,
			true,
		);
	}

	warning(message: string, name?: string): this {
		return this.add(
			() => {
				warning(message);

				return null;
			},
			name,
			true,
		);
	}

	alert(message: string, name?: string): this {
		return this.add(
			() => {
				alert(message);

				return null;
			},
			name,
			true,
		);
	}

	info(message: string, name?: string): this {
		return this.add(
			() => {
				info(message);

				return null;
			},
			name,
			true,
		);
	}

	intro(message: string, name?: string): this {
		return this.add(
			() => {
				intro(message);

				return null;
			},
			name,
			true,
		);
	}

	outro(message: string, name?: string): this {
		return this.add(
			() => {
				outro(message);

				return null;
			},
			name,
			true,
		);
	}

	table(headersOrOptions: TableOptions | string[] = [], rows: TableOptions['rows'] | null = null, name?: string): this {
		return this.add(
			() => {
				table(headersOrOptions, rows);

				return null;
			},
			name,
			true,
		);
	}

	datatable<T = unknown>(options: DataTablePromptOptions<T>, name?: string): this {
		return this.add(() => datatable(options), name);
	}

	progress<T, R>(label: string, steps: Iterable<T> | number, callback?: (step: T | number, bar: Progress) => MaybePromise<R>, hint = '', name?: string): this {
		return this.add(() => progress(label, steps, callback, hint), name, true);
	}
}

File: provision/go/internal/form/builder.go
package form

import "errors"

type StepFunc func(responses *Responses, previous any, name string) (any, error)
type ConditionFunc func(responses *Responses) (bool, error)

type Step struct {
	Condition           ConditionFunc
	IgnoreWhenReverting bool
	Name                string
	Run                 StepFunc
}

type Builder struct {
	responses *Responses
	steps     []Step
}

func New() *Builder {
	return &Builder{responses: NewResponses()}
}

func (builder *Builder) Add(run StepFunc, name ...string) *Builder {
	return builder.addStep(trueCondition, run, firstName(name), false)
}

func (builder *Builder) AddSideEffect(run func() error, name ...string) *Builder {
	return builder.addStep(trueCondition, SideEffect(run), firstName(name), true)
}

func (builder *Builder) AddIf(condition ConditionFunc, run StepFunc, name ...string) *Builder {
	return builder.addStep(condition, run, firstName(name), false)
}

func (builder *Builder) AddSideEffectIf(condition ConditionFunc, run func() error, name ...string) *Builder {
	return builder.addStep(condition, SideEffect(run), firstName(name), true)
}

func (builder *Builder) Submit() (*Responses, error) {
	index := 0
	wasReverted := false

	for index < len(builder.steps) {
		step := builder.steps[index]

		if wasReverted && index > 0 {
			ignore, err := shouldIgnoreWhenReverting(step, builder.responses)
			if err != nil {
				return nil, err
			}
			if ignore {
				index--
				continue
			}
		}

		wasReverted = false
		shouldRun, err := shouldRun(step, builder.responses)
		if err != nil {
			return nil, err
		}
		if !shouldRun {
			builder.responses.set(index, step.Name, nil)
			index++
			continue
		}

		previous := builder.responses.previous(index, step.Name)
		value, err := runWithRevert(index > 0, func() (any, error) {
			return step.Run(builder.responses, previous, step.Name)
		})
		if err != nil {
			if !errors.Is(err, ErrReverted) {
				return nil, err
			}
			wasReverted = true
		} else {
			builder.responses.set(index, step.Name, value)
		}

		if wasReverted {
			index--
		} else {
			index++
		}
	}

	return builder.responses, nil
}

func (builder *Builder) addStep(condition ConditionFunc, run StepFunc, name string, ignoreWhenReverting bool) *Builder {
	if condition == nil {
		condition = trueCondition
	}
	builder.steps = append(builder.steps, Step{
		Condition:           condition,
		IgnoreWhenReverting: ignoreWhenReverting,
		Name:                name,
		Run:                 run,
	})
	return builder
}

func SideEffect(callback func() error) StepFunc {
	return func(*Responses, any, string) (any, error) {
		if callback == nil {
			return nil, nil
		}

		return nil, callback()
	}
}

func Bool(condition bool) ConditionFunc {
	return func(*Responses) (bool, error) {
		return condition, nil
	}
}

func trueCondition(*Responses) (bool, error) {
	return true, nil
}

func shouldRun(step Step, responses *Responses) (bool, error) {
	if step.Condition == nil {
		return true, nil
	}

	return step.Condition(responses)
}

func shouldIgnoreWhenReverting(step Step, responses *Responses) (bool, error) {
	run, err := shouldRun(step, responses)
	if err != nil || !run {
		return true, err
	}

	return step.IgnoreWhenReverting, nil
}

func firstName(values []string) string {
	if len(values) == 0 {
		return ""
	}

	return values[0]
}
